subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt

```diff
@@ -322,18 +327,24 @@ class DefaultConfigurationCache internal constructor(
 
     private
     fun saveModel(model: Any) {
-        saveToCache(StateType.Model) { stateFile ->
-            cacheIO.writeModelTo(model, stateFile)
-        }
+        saveToCache(
+            stateType = StateType.Model,
+            action = { stateFile -> cacheIO.writeModelTo(model, stateFile) },
+            onSaveFinish = { scopeRegistryListener.dispose() }
+        )
     }
 
     private
-    fun saveWorkGraph() {
-        saveToCache(StateType.Work) { layout -> writeConfigurationCacheState(layout) }
+    fun saveWorkGraph(isModelBuildingRequested: Boolean) {
+        saveToCache(
+            stateType = StateType.Work,
+            action = { stateFile -> writeConfigurationCacheState(stateFile) },
+            onSaveFinish = { if (!isModelBuildingRequested) scopeRegistryListener.dispose() }
+        )
     }
 
     private
-    fun saveToCache(stateType: StateType, action: (ConfigurationCacheStateFile) -> Unit) {
+    fun saveToCache(stateType: StateType, action: (ConfigurationCacheStateFile) -> Unit, onSaveFinish: () -> Unit) {
```

Comment:
💅 I'd pass the boolean down to this method (as something like `disposeResources`) so that the `scopeRegistryListener.dispose()` action is mentioned only once.